### PR TITLE
SCC API check: server

### PR DESCRIPTION
**Pull request by sdminonne**
@deads2k

### CODE REVIEW

File: hack/common.sh [hunk 1]
```diff
@@ -663,7 +663,7 @@ function os::build::os_version_vars() {
       # Try to match the "git describe" output to a regex to try to extract
       # the "major" and "minor" versions and whether this is the exact tagged
       # version or whether the tree is between two tagged versions.
-      if [[ "${OS_GIT_VERSION}" =~ ^v([0-9]+)\.([0-9]+)(\.[0-9]+)?([-].*)?$ ]]; then
+      if [[ "${OS_GIT_VERSION}" =~ ^v([0-9]+)\.([0-9]+)(\.[0-9]+)?([-,+].*)?$ ]]; then
```
Comment (stevekuznetsov) [inline hack/common.sh]: @smarterclayton PTAL this

Comment (smarterclayton) [inline hack/common.sh]: Where did the + come from?

Comment (sdminonne) [inline hack/common.sh]: @smarterclayton my fault I didn't want to commit and I forgot to revert this change.
In my clone `$OS_GIT_VERSION` is `v1.0.6+f1a2c89-dirty` and it doesn't match the regex.
I didn't have time to  debug  so I modified the regex... The intent was not to commit.

File: pkg/security/registry/podsecuritypolicyreview/rest.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// Create registers a given new PodSecurityPolicyReview instance to r.registry.
+func (r *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, error) {
+	pspr, ok := obj.(*securityapi.PodSecurityPolicyReview)
+	if !ok {
+		return nil, kapierrors.NewBadRequest(fmt.Sprintf("not a PodSecurityPolicyReview: %#v", obj))
+	}
+	if errs := securityvalidation.ValidatePodSecurityPolicyReview(pspr); len(errs) > 0 {
+		return nil, kapierrors.NewInvalid(securityapi.Kind(pspr.Kind), "", errs)
```
Comment (deads2k) [inline pkg/security/registry/podsecuritypolicyreview/rest.go]: You'll probably have an empty kind here.  Use the plain string for now.

File: pkg/security/registry/podsecuritypolicyreview/rest.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	if len(serviceAccounts) == 0 {
+		glog.Errorf("No service accounts for namespace %s", ns)
+		return pspr, nil
+	}
+
+	errs := []error{}
+	newStatus := securityapi.PodSecurityPolicyReviewStatus{}
+	sccMatcher := oscc.NewDefaultSCCMatcher(r.sccLister)
```
Comment (deads2k) [inline pkg/security/registry/podsecuritypolicyreview/rest.go]: At first glance, this interface looks threadsafe.  Can the `REST` simply hold the `SCCMatcher`?  Maybe even have it passed in for easier mocking?

File: pkg/security/registry/podsecuritypolicyreview/rest.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	if errs := securityvalidation.ValidatePodSecurityPolicyReview(pspr); len(errs) > 0 {
+		return nil, kapierrors.NewInvalid(securityapi.Kind(pspr.Kind), "", errs)
+	}
+	ns, ok := kapi.NamespaceFrom(ctx)
+	if !ok {
+		return pspr, kapierrors.NewBadRequest("namespace parameter required.")
+	}
+
+	serviceAccounts, err := getServiceAccounts(pspr.Spec, r.client, ns)
```
Comment (deads2k) [inline pkg/security/registry/podsecuritypolicyreview/rest.go]: we ought to have a serviceaccount lister by now.

Comment (deads2k) [inline pkg/security/registry/podsecuritypolicyreview/rest.go]: > we ought to have a serviceaccount lister by now.

wow, we don't.  I'll add it.

File: pkg/security/registry/podsecuritypolicyreview/rest.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+		oscc.DeduplicateSecurityContextConstraints(saConstraints)
+		sort.Sort(oscc.ByPriority(saConstraints))
+		var namespace *kapi.Namespace
+		for _, constraint := range saConstraints {
+			var (
+				provider kscc.SecurityContextConstraintsProvider
+				err      error
+			)
+			pspsrs := securityapi.PodSecurityPolicySubjectReviewStatus{}
+			if provider, namespace, err = oscc.CreateProviderFromConstraint(ns, namespace, constraint, r.client); err != nil {
```
Comment (deads2k) [inline pkg/security/registry/podsecuritypolicyreview/rest.go]: I think this logic (deduping, sorting, creating provider, pod manipulation) should all be abstracted.

**Comment by deads2k** [comment]
The intent looks correct, but I think we'll need to split this into a refactor commit or two and then the `RESTStorage`.  It's hard for me to review the moves with everything else.  How about we split the refactors and do those first, post-3.3?

I won't have time to review this in detail until after 3.3.

**Comment by sdminonne** [comment]
@deads2k OK, no problem. I'm going to split this in 3 PRs:

- move 
- refactor (after move)
- REST

**Comment by deads2k** [comment]
@mfojtik this is for one of the cards.

**Comment by soltysh** [comment]
This was split into following PRs:
- move - https://github.com/openshift/origin/pull/10631
- refactor - https://github.com/openshift/origin/pull/10780
- RESTStorage - TBD
Can we then close this one?
